rust/src/push_consumer.rs

```diff
@@ -0,0 +1,1303 @@
+        }
+    }
+
+    #[allow(clippy::needless_lifetimes)]
+    pub(crate) fn get_filter_expression<'a>(&'a self) -> Option<&'a FilterExpression> {
+        self.option
+            .subscription_expressions()
+            .get(self.message_queue.topic.name.as_str())
+    }
+
+    pub(crate) async fn start(&mut self) -> Result<(), ClientError> {
+        debug!(
+            self.logger,
+            "start a new queue actor {:?}", self.message_queue
+        );
+        let (tx, mut shutdown_rx) = oneshot::channel();
+        self.shutdown_tx = Some(tx);
+        let (tx, mut poll_rx) = mpsc::channel(8);
+        let poll_tx: mpsc::Sender<()> = tx;
+        let logger = self.logger.clone();
+        let mut actor = self.fork_self();
+        let mut message_handler_queue: VecDeque<MessageHandler> = VecDeque::new();
+        let (tx, mut ack_rx) = mpsc::channel(8);
+        let ack_tx: mpsc::Sender<()> = tx;
+        let mut queue_check_ticker = tokio::time::interval(std::time::Duration::from_secs(1));
+        let mut rpc_client = actor.rpc_client.shadow_session();
+        tokio::spawn(async move {
+            loop {
+                select! {
+                    _ = poll_rx.recv() => {
+                        actor.receive_messages(&mut rpc_client, &mut message_handler_queue, poll_tx.clone()).await;
+                    }
+                    _ = ack_rx.recv() => {
+                        actor.ack_message_in_waiting_queue(&mut rpc_client, &mut message_handler_queue, ack_tx.clone()).await;
+                    }
+                    _ = queue_check_ticker.tick() => {
+                        if !message_handler_queue.is_empty() {
+                            let _ = ack_tx.try_send(());
+                        }
+                        if message_handler_queue.len() < actor.option.max_cache_message_count() {
+                            let _ = poll_tx.try_send(());
+                        }
+                    }
+                    _ = &mut shutdown_rx => {
+                        info!(logger, "message queue actor shutdown");
+                        actor.flush_waiting_queue(&mut rpc_client, &mut message_handler_queue).await;
+                        break;
+                    }
+                }
+            }
+        });
+        Ok(())
+    }
+
+    async fn receive_messages<T: RPCClient + 'static>(
+        &mut self,
+        rpc_client: &mut T,
+        message_handler_queue: &mut VecDeque<MessageHandler>,
+        poll_tx: mpsc::Sender<()>,
+    ) {
+        let poll_result = self.poll_messages(rpc_client).await;
+        if let Ok(messages) = poll_result {
+            for message in messages {
+                let consume_result = (self.message_listener)(&message);
+                match consume_result {
+                    ConsumeResult::SUCCESS => {
+                        let result = self.ack_message(rpc_client, &message).await;
+                        if result.is_err() {
+                            info!(
```

Comment:
It would be better to define an abstract model here, some kind of ProcessTask and encapsulate process-ack-retry and order-keeping.
That is, an message queue actor has one(FIFO) or multiple(standard model) ProcessTask, one message arrival, put messages to existing ProcessTask(for FIFO) or spawn new ProcessTasks.

In the process task, we have the consume semantics implemented.

Reply:
You idea is great, and I refactored that the consumption of standard and fifo is capsulated in separate implementation. And I also extract the acknowledgement work into a single implementation.